File: src/algorithms/graph/shortest-path/dijkstra/sources/Dijkstra_test.java
import java.util.*;

// Compile: javac Dijkstra.java Dijkstra_test.java
// Run:     java -ea Dijkstra_test
public class Dijkstra_test {
    public static void main(String[] args) {
        testComputesShortestDistancesInSimpleWeightedGraph();
        testReturnsZeroDistanceForStartNode();
        testReturnsInfinityForUnreachableNodes();
        testHandlesSingleNodeGraph();
        testFindsShortestPathThroughMultipleHops();
        testUsesLowerWeightPathOverDirectPath();
        testHandlesLinearChainCorrectly();
        testHandlesEqualWeightEdges();
        System.out.println("All tests passed!");
    }

    static Map<String, List<Object[]>> adj(Object[]... entries) {
        Map<String, List<Object[]>> map = new LinkedHashMap<>();
        for (Object[] entry : entries) {
            String node = (String) entry[0];
            List<Object[]> neighbors = new ArrayList<>();
            for (int edgeIdx = 1; edgeIdx < entry.length; edgeIdx++) {
                neighbors.add((Object[]) entry[edgeIdx]);
            }
            map.put(node, neighbors);
        }
        return map;
    }

    static void testComputesShortestDistancesInSimpleWeightedGraph() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"A", new Object[]{"B", 4}, new Object[]{"C", 2}},
            new Object[]{"B", new Object[]{"D", 5}},
            new Object[]{"C", new Object[]{"B", 1}, new Object[]{"D", 8}},
            new Object[]{"D"}
        );
        Map<String, Double> d = Dijkstra.dijkstraShortestPath(adjacencyList, "A");
        assert d.get("A") == 0.0;
        assert d.get("B") == 3.0;
        assert d.get("C") == 2.0;
        assert d.get("D") == 8.0;
    }

    static void testReturnsZeroDistanceForStartNode() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"X", new Object[]{"Y", 10}},
            new Object[]{"Y"}
        );
        Map<String, Double> d = Dijkstra.dijkstraShortestPath(adjacencyList, "X");
        assert d.get("X") == 0.0;
    }

    static void testReturnsInfinityForUnreachableNodes() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"A", new Object[]{"B", 1}},
            new Object[]{"B"},
            new Object[]{"C"}
        );
        Map<String, Double> d = Dijkstra.dijkstraShortestPath(adjacencyList, "A");
        assert d.get("C") == Double.MAX_VALUE || d.get("C").isInfinite();
    }

    static void testHandlesSingleNodeGraph() {
        Map<String, List<Object[]>> adjacencyList = adj(new Object[]{"A"});
        Map<String, Double> d = Dijkstra.dijkstraShortestPath(adjacencyList, "A");
        assert d.get("A") == 0.0;
    }

    static void testFindsShortestPathThroughMultipleHops() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"A", new Object[]{"B", 4}, new Object[]{"C", 2}},
            new Object[]{"B", new Object[]{"D", 5}},
            new Object[]{"C", new Object[]{"B", 1}, new Object[]{"D", 8}, new Object[]{"E", 10}},
            new Object[]{"D", new Object[]{"F", 2}},
            new Object[]{"E", new Object[]{"F", 3}},
            new Object[]{"F"}
        );
        Map<String, Double> d = Dijkstra.dijkstraShortestPath(adjacencyList, "A");
        assert d.get("C") == 2.0;
        assert d.get("B") == 3.0;
        assert d.get("D") == 8.0;
        assert d.get("F") == 10.0;
        assert d.get("E") == 12.0;
    }

    static void testUsesLowerWeightPathOverDirectPath() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"A", new Object[]{"B", 10}, new Object[]{"C", 1}},
            new Object[]{"B", new Object[]{"D", 1}},
            new Object[]{"C", new Object[]{"B", 1}, new Object[]{"D", 5}},
            new Object[]{"D"}
        );
        Map<String, Double> d = Dijkstra.dijkstraShortestPath(adjacencyList, "A");
        assert d.get("D") == 3.0;
    }

    static void testHandlesLinearChainCorrectly() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"A", new Object[]{"B", 2}},
            new Object[]{"B", new Object[]{"C", 3}},
            new Object[]{"C", new Object[]{"D", 4}},
            new Object[]{"D"}
        );
        Map<String, Double> d = Dijkstra.dijkstraShortestPath(adjacencyList, "A");
        assert d.get("B") == 2.0;
        assert d.get("C") == 5.0;
        assert d.get("D") == 9.0;
    }

    static void testHandlesEqualWeightEdges() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"A", new Object[]{"B", 1}, new Object[]{"C", 1}},
            new Object[]{"B", new Object[]{"D", 1}},
            new Object[]{"C", new Object[]{"D", 1}},
            new Object[]{"D"}
        );
        Map<String, Double> d = Dijkstra.dijkstraShortestPath(adjacencyList, "A");
        assert d.get("D") == 2.0;
    }
}

File: src/algorithms/graph/shortest-path/dijkstra/__tests__/dijkstra_test.py
import importlib
import sys
import os
import math

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "sources"))
module = importlib.import_module("dijkstra")
dijkstra_shortest_path = module.dijkstra_shortest_path


def test_computes_shortest_distances_in_simple_weighted_graph():
    adj = {
        "A": [("B", 4), ("C", 2)],
        "B": [("D", 5)],
        "C": [("B", 1), ("D", 8)],
        "D": [],
    }
    distances = dijkstra_shortest_path(adj, "A")
    assert distances["A"] == 0
    assert distances["B"] == 3
    assert distances["C"] == 2
    assert distances["D"] == 8


def test_returns_zero_distance_for_start_node():
    adj = {"X": [("Y", 10)], "Y": []}
    distances = dijkstra_shortest_path(adj, "X")
    assert distances["X"] == 0


def test_returns_infinity_for_unreachable_nodes():
    adj = {"A": [("B", 1)], "B": [], "C": []}
    distances = dijkstra_shortest_path(adj, "A")
    assert math.isinf(distances["C"])


def test_handles_single_node_graph():
    adj = {"A": []}
    distances = dijkstra_shortest_path(adj, "A")
    assert distances["A"] == 0


def test_finds_shortest_path_through_multiple_hops():
    adj = {
        "A": [("B", 4), ("C", 2)],
        "B": [("D", 5)],
        "C": [("B", 1), ("D", 8), ("E", 10)],
        "D": [("F", 2)],
        "E": [("F", 3)],
        "F": [],
    }
    distances = dijkstra_shortest_path(adj, "A")
    assert distances["A"] == 0
    assert distances["C"] == 2
    assert distances["B"] == 3
    assert distances["D"] == 8
    assert distances["F"] == 10
    assert distances["E"] == 12


def test_uses_lower_weight_path_over_direct_path():
    adj = {
        "A": [("B", 10), ("C", 1)],
        "B": [("D", 1)],
        "C": [("B", 1), ("D", 5)],
        "D": [],
    }
    distances = dijkstra_shortest_path(adj, "A")
    assert distances["D"] == 3


def test_handles_linear_chain_correctly():
    adj = {"A": [("B", 2)], "B": [("C", 3)], "C": [("D", 4)], "D": []}
    distances = dijkstra_shortest_path(adj, "A")
    assert distances["B"] == 2
    assert distances["C"] == 5
    assert distances["D"] == 9


def test_handles_equal_weight_edges():
    adj = {"A": [("B", 1), ("C", 1)], "B": [("D", 1)], "C": [("D", 1)], "D": []}
    distances = dijkstra_shortest_path(adj, "A")
    assert distances["D"] == 2


if __name__ == "__main__":
    test_computes_shortest_distances_in_simple_weighted_graph()
    test_returns_zero_distance_for_start_node()
    test_returns_infinity_for_unreachable_nodes()
    test_handles_single_node_graph()
    test_finds_shortest_path_through_multiple_hops()
    test_uses_lower_weight_path_over_direct_path()
    test_handles_linear_chain_correctly()
    test_handles_equal_weight_edges()
    print("All tests passed!")
